Show ai content as regular blocks

n3f opened the pull request:
Fixes #28384

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Remove the content attribute for the ai-paragraph and use a paragraph block for storing the returned data from openAI.

## Testing instructions:
1. Apply to your sandbox with `bin/jetpack-downloader test jetpack ai/add-editable-paragraph`.
2. Load your sandboxed site and test the ai-paragraph block as outlined below.

## Does this pull request change what data or activity we track or use?

No

## Things to test:
- [x] Try creating an open ai block with no content (no title, no tags, no categories). You should see an error message without a retry box.  After you've added some content the retry should appear.
- [x] Force the block into an error state by using the code editor and changing state to "retry". This error state adds a generic message and shows the retry button.  Using "error" as state will not show the retry button.
- [x] Open an old ai-paragraph block.  The migration should work. (The new block will have a single `InnerBlock` of a paragraph). An example old ai-paragraph block can be found here: 2f768-pb - This can be pasted into the code editor.
- [x] Create a new ai-paragraph block. After it gets created you should be able to edit it like a normal post.
- [x] After the new block is created, the preview should display correctly.
- [x] The AI block now transforms itself into the group block correctly (just passes the `innerBlocks` into a new group block).

n3f committed:
[not verified] Add editable paragraph to ai paragraph.

n3f committed:
[not verified] changelog

n3f commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -278,6 +299,31 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
+		}, 50 * tokens.length );
+	}, [ content ] );
+
+	const TEMPLATE = [
+		[ 'core/paragraph', {
+			placeholder: ' ',
+		} ]
+	];
```
placeholder won't be needed once I add a CSS fix to hide the paragraph while it's loading.

n3f commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -290,17 +336,18 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
 
-			{ contentIsLoaded && (
+			{/* { contentIsLoaded && (
 				<ShowLittleByLittle
 					showAnimation={ ! attributes.animationDone }
 					onAnimationDone={ () => {
+						updateInnerBlocks( clientId, attributes.content );
 						setAttributes( { animationDone: true } );
 					} }
 					html={ attributes.content }
 				/>
-			) }
+			) } */}
```
TODO: remove

n3f commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -278,6 +299,31 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
+	];
+
 	return (
 		<div { ...useBlockProps() }>
```
Add a loading class to hide the innerblocks...

n3f commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -70,6 +70,25 @@ const createPrompt = ( postTitle, contentBeforeCurrentBlock, tagsAndCategoriesNa
+};
+
 // This component displays the text word by word if show animation is true
 function ShowLittleByLittle( { html, showAnimation, onAnimationDone } ) {
```
TODO: Remove

n3f commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -278,6 +299,31 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
+			return;
+		}
+
+		console.log( 'contentIsLoaded', content );
```
remove

n3f commented:
@millerf, what do you think?  Can I simplify some of the logic more now?

n3f commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -278,6 +299,31 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
 		}
 	}
 
+	useEffect( ( c1 ) => {
```
`c1` not needed -- I was testing something.

millerf commented:
@n3f I briefly tested (without looking at the code), and it seems to work, however converting it to paragraph doesn't work anymore:

https://user-images.githubusercontent.com/790558/214847514-d11ef53e-6ae3-4f45-96c0-078fd9927c80.mov

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -232,9 +253,9 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
+	const contentIsLoaded = !! content;
 
-	// We do nothing is we are waiting for stuff OR if the content is already loaded.
+	// We do nothing if we are waiting for stuff OR if the content is already loaded.
```
Oups.

millerf commented:
@n3f The approach seems like a good one. But there are a couple of issues to iron out:
- See [this comment](https://github.com/Automattic/jetpack/pull/28580#issuecomment-1405009781). But we could remove the "transform" option from `index.js`
- Reloading the page keeps the paragraph but triggers a new call to the back-end...

https://user-images.githubusercontent.com/790558/214870414-2a039a5b-378d-47de-8513-f8928893db49.mov


Finally, I stlll struggle to deal with innerblocks. Like deleting the parent block becomes a chore...
Could we instead extend the paragraph block? I have no idea if it is possible, I am just thinking out loud...

romarioraffington commented:
Noting here that I think the InnerBlocks approach appears to a bit more complex than the `RichText` approach in https://github.com/Automattic/jetpack/pull/28575. FWIW, I tried replicating the issues Fab raised in https://github.com/Automattic/jetpack/pull/28575 but couldn't reproduce there.

I have no strong opinions on which approach we should move forward with.

I added a **Possible Regression / Improvements** section in https://github.com/Automattic/jetpack/pull/28575 and I think that section also applies here. @artpi  do you have any strong opinions?

n3f committed:
[not verified] Various fixes.

n3f committed:
[not verified] Use the markdown block instead!

n3f commented:
Easy to support different blocks! Now it supports markdown!

n3f committed:
[not verified] Add comments and fix showing block after content loaded.

n3f committed:
[not verified] Try fixing some linting.

n3f committed:
[not verified] Use `pasteHandler` to convert HTML into blocks.

n3f committed:
[not verified] Use gutenberg markdown processor.

n3f committed:
[not verified] Rebase & small fixes.

n3f committed:
[not verified] Functionally working!

n3f committed:
[not verified] Add deprecation/migration.

n3f committed:
[not verified] Transform to group

millerf commented:
So... Many... Labels...

<img width="245" alt="Screenshot 2023-03-09 at 11 09 43" src="https://user-images.githubusercontent.com/790558/223914574-1cfcc5bf-c7c2-4e77-9997-df030c5356d1.png">

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/deprecated/v1.js:
````diff
@@ -0,0 +1,45 @@
+ * Example data to migrate:
+ * ```
+ * <!-- wp:paragraph -->
+ * <p>GPT generates a block for testing.  This block should be an interesting but somewhat obscure fact regarding pirates and their constant battles with ninjas.</p>
````
Lol?

n3f replied:
I could add some more comments for this.  But essentially if you open the code editor, & paste this in it will try to convert the blocks to the new version.

n3f replied:
Prettier doesn't like me making it easy to copy/paste so I'm just going to remove.

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/deprecated/v1.js:
````diff
@@ -0,0 +1,45 @@
+ * <!-- /wp:jetpack/ai-paragraph -->
+ * ```
+ */
+export default {
````
Any idea how to test that?

n3f replied:
See above.

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -2,16 +2,22 @@ import './editor.scss';
 export const MAXIMUM_NUMBER_OF_CHARACTERS_SENT_FROM_CONTENT = 240;
 
+const is_untriggered = state => state === AI_BLOCK_STATE.DEFAULT;
+const is_triggered = state => state !== AI_BLOCK_STATE.DEFAULT;
```
Eerrrrr... `const is_triggered = (s) => !is_untriggered(s)`? Do we really need both constants? :)

n3f replied:
Probably not, but as it was one line and it made the code more readable I prefer it.  Also, there is a possibility that the logic might change. (For instance the untriggered state might actually need to be none of the defined states -- I saw a weird state happen when I was testing deprecations where this would have been a good fix)

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -201,14 +174,17 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
 			.then( res => {
-				const result = res.prompts[ 0 ].text.trim().replaceAll( '\n', '<br/>' );
-				setAttributes( { content: result } );
+				const result = res.prompts[ 0 ].text.trim();
```
interesting...

n3f replied:
The entire result just gets passed to `pasteHandler`.

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -274,8 +250,59 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
+	);
+
+	// This is to animate text input. This will give an idea of a "better" AI.
+	// At this point this is an established pattern.
```
:)

n3f replied:
😆 - just marking this was an existing comment.

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -274,8 +250,59 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
 
+	const innerBlocks = useSelect( select => {
+		const aiParagraphBlock = select( 'core/block-editor' ).getBlocksByClientId( clientId );
+		return aiParagraphBlock[ 0 ].innerBlocks;
```
OI am unsure, is that always the first paragraph?

n3f replied:
It's not the first paragraph, it's the actual 'AI paragraph block'.  Maybe we should rename the block with this PR? (AI text)

millerf replied:
Oh, I see 👍

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -274,8 +250,59 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
+
+	const { replaceInnerBlocks } = useDispatch( 'core/block-editor' );
+
+	const updateInnerBlocks = useCallback(
```
I struggle to understand the use of useCallback here... That is definitely from a lack of knowledge for me... Can you ad a comment?

n3f replied:
It's not needed per se, all `useCallback` calls are for performance reasons (or for getting re-used in other useEffect callbacks).

millerf commented on projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js:
```diff
@@ -274,8 +250,59 @@ export default function Edit( { attributes, setAttributes, clientId } ) {
 	return (
-		<div { ...useBlockProps() }>
+		<div { ...blockProps }>
+			<InnerBlocks />
```
This element should be hidden in the PROCESSING state...

n3f replied:
Yup, I had that at one point and I took it out yesterday because I couldn't remember why I had it 😆 (Hidden by CSS).

millerf reviewed (dismissed):
The global logic looks solid. There is still some polishing to be done but it is very close!

n3f committed:
[not verified] Rebase & review fixes.

romarioraffington reviewed (dismissed):
Tests well and code LGTM

jeherve reviewed (commented):
This doesn't seem to work for me; I get errors when opening old posts with AI paragraph blocks:

<img width="1045" alt="image" src="https://user-images.githubusercontent.com/426388/225291905-9011ab43-e1be-4ecc-b338-3c860c1d37af.png">

romarioraffington commented:
I got the same error when testing at first but couldn't reproduce once it: Discussion in p1678750861908199-slack-C01B6KEJ5GE

n3f commented:
> This doesn't seem to work for me; I get errors when opening old posts with AI paragraph blocks:

@jeherve, do you have the copy that you are using for the existing post?

jeherve commented:
> do you have the copy that you are using for the existing post?

Sure thing, I'll send you the info in slack.

n3f committed:
[not verified] Use RawHTML for deprecated v1 save function.

n3f commented:
I wasn't able to get the same error message, but I did see a problem with some v1 content as the `save()` function wasn't identical to the previous save function.  I fixed this and I was unable to trigger any other errors.

* I tested by using the current version of the ai-paragraph block and creating a half-dozen uses of it (published, drafts, with categories, without categories, etc.)
* Then I reloaded the code from this PR, and re-opened the post editors for each one.
* Pretty sure it failed when it came to a `<br/>` in a post. (The post that triggered this was a no-categories draft, but I don't think that was actually relevant).

@jeherve should I have @Automattic/gold take a look at this again, or do you want to give it a shot?

jeherve commented:
I keep seeing the error when trying to test, even with new posts created with the new version of the block :\

It's probably best if I let someone else test this.

n3f committed:
[not verified] Various state fixes.

- Moved state props to own file.
- Made block state more observable, added comments and hopefully helpful hints.
- Simplified JSX.

n3f committed:
[not verified] Fix example.

romarioraffington reviewed (dismissed):
>Apply to your sandbox with bin/jetpack-downloader test jetpack ai/add-editable-paragraph.
Load your sandboxed site and test the ai-paragraph block as outlined below.

Had to change the `JETPACK_PLUGIN_LOADER` to staging to see the changes 🤔 

> Open an old ai-paragraph block. The migration should work. (The new block will have a single InnerBlock of a paragraph). An example old ai-paragraph block can be found here: 2f768-pb - This can be pasted into the code editor.I 

I tested this by using the production version of the blocks to generate content before testing your PR. It WFM

millerf reviewed (dismissed):
LGTM and works like a charm :shipit: after fixing the tests issues

millerf commented:
@n3f This one also fixes https://github.com/Automattic/jetpack/issues/29381, right?

n3f commented:
> @n3f This one also fixes #29381, right?

Yup!

> LGTM and works like a charm :shipit: after fixing the tests issues.

I'm not sure what the test issues are.  There shouldn't be any problems with this...  (It looks to me like flaky tests, but if you know something I don't I'm all ears.)

n3f committed:
Fix linting.

jeherve commented:
Reviewing myself from the review request. I still cannot get this to work on my sandbox, so my review would not be complete. I'll let somebody else give this a try. 👍 

fwiw I cannot reproduce the error I previously ran into here https://github.com/Automattic/jetpack/pull/28580#pullrequestreview-1341223463
I am now able to add new blocks to new posts, but when editing existing posts with an AI block, the content never finishes loading, it just spins forever.

n3f commented:
> fwiw I cannot reproduce the error I previously ran into here [#28580 (review)](https://github.com/Automattic/jetpack/pull/28580#pullrequestreview-1341223463) I am now able to add new blocks to new posts, but when editing existing posts with an AI block, the content never finishes loading, it just spins forever.

Anything in the dev console?

I have been seeing a longer response time too, but nothing past 30 seconds or so...

jeherve commented:
> Anything in the dev console?

Nothing, no, even after a few minutes.
